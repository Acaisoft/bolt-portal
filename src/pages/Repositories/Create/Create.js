import React, { Component } from 'react'
import PropTypes from 'prop-types'

import { Typography } from '@material-ui/core'
import { RepositoryForm } from '~containers/forms'

export class Create extends Component {
  static propTypes = {}

  render() {
    return (
      <div>
        <Typography variant="h6">Repository Data</Typography>
        <Typography variant="body1">
          Here you can define your new test run repository, you can edit the data
          until first test run. After that you will be able only to change
          configuration name.
        </Typography>

        <RepositoryForm mode="create" />
      </div>
    )
  }
}

export default Create
