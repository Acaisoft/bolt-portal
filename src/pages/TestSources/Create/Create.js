import React, { useCallback } from 'react'
import PropTypes from 'prop-types'

import { useQuery } from 'react-apollo-hooks'
import {
  Button,
  MenuItem,
  Grid,
  withStyles,
  FormHelperText,
} from '@material-ui/core'
import { Add, VpnKey } from '@material-ui/icons'
import { Field } from 'react-final-form'
import { TestSourceForm } from '~containers/forms'
import { FormCondition, FormField } from '~containers'
import { ButtonWithIcon, Loader, SectionHeader, CopyToClipboard } from '~components'

import { TestSourceType } from '~config/constants'
import routes from '~config/routes'
import { getUrl } from '~utils/router'
import { useToggle } from '~hooks'

import { GET_REPOSITORY_KEY } from './graphql'
import styles from './Create.styles'

export function Create({ classes, history, match }) {
  const [isKeyVisible, toggleKeyInput] = useToggle(false)

  const {
    data: { repositoryKey },
    loading,
  } = useQuery(GET_REPOSITORY_KEY, {
    fetchPolicy: 'cache-first',
  })

  const handleSubmit = useCallback(
    values => {
      history.push(getUrl(routes.projects.sources.list, { ...match.params }))
    },
    [history, match]
  )

  const handleCancel = useCallback(() => {
    history.goBack()
  }, [history])

  const { projectId, testSourceId } = match.params

  const mode = testSourceId ? 'edit' : 'create'

  if (loading) {
    return <Loader loading />
  }

  return (
    <div>
      <TestSourceForm mode="create" projectId={projectId} onSubmit={handleSubmit}>
        {({ form, fields, handlers, isTestingConnection, isConnectionOk }) => {
          const repositoryFields = fields.repository.fields

          return (
            <React.Fragment>
              <SectionHeader title="Test Source">
                <Button
                  color="default"
                  variant="text"
                  disabled={form.submitting}
                  onClick={handleCancel}
                >
                  Cancel
                </Button>
                <ButtonWithIcon
                  className={classes.marginLeft}
                  color="secondary"
                  variant="contained"
                  disabled={
                    !isConnectionOk ||
                    !form.dirty ||
                    form.submitting ||
                    form.validating ||
                    form.invalid ||
                    form.dirtySinceLastSubmit
                  }
                  icon={Add}
                  onClick={form.handleSubmit}
                >
                  {mode === 'create' ? 'Create' : 'Update'}
                </ButtonWithIcon>
              </SectionHeader>

              <form onSubmit={form.handleSubmit}>
                <Grid container spacing={32}>
                  <Grid item xs={12} md={6}>
                    <FormField
                      name="source_type"
                      field={fields.source_type}
                      margin="normal"
                      variant="filled"
                      fullWidth
                    >
                      {fields.source_type.options.map(type => (
                        <MenuItem
                          key={type.key}
                          value={type.value}
                          disabled={type.value === TestSourceType.TEST_CREATOR}
                        >
                          {type.label}
                        </MenuItem>
                      ))}
                    </FormField>
                  </Grid>
                  <Grid item xs={12} md={6} />

                  <FormCondition when="source_type" is={TestSourceType.REPOSITORY}>
                    <Grid item xs={12} md={6}>
                      <FormField
                        name="repository.name"
                        field={repositoryFields.name}
                        margin="normal"
                        variant="filled"
                        fullWidth
                      />
                    </Grid>
                    <Grid item xs={12} md={6}>
                      <FormField
                        name="repository.type_slug"
                        field={repositoryFields.type_slug}
                        margin="normal"
                        variant="filled"
                        fullWidth
                      >
                        {repositoryFields.type_slug.options.map(type => (
                          <MenuItem key={type.key} value={type.value}>
                            {type.label}
                          </MenuItem>
                        ))}
                      </FormField>
                    </Grid>
                    <Grid item xs={12} md={6}>
                      <FormField
                        name="repository.url"
                        field={repositoryFields.url}
                        margin="normal"
                        variant="filled"
                        fullWidth
                      />
                      <div>
                        {isKeyVisible && (
                          <CopyToClipboard
                            text={repositoryKey}
                            label="Repository Key"
                            margin="normal"
                            variant="filled"
                            multiline
                            fullWidth
                          />
                        )}
                        <ButtonWithIcon
                          color="primary"
                          variant="contained"
                          onClick={() => toggleKeyInput()}
                          icon={VpnKey}
                        >
                          {isKeyVisible ? 'Hide' : 'Show'} Key
                        </ButtonWithIcon>
                        <Button
                          color="primary"
                          variant="contained"
                          disabled={!form.valid || isTestingConnection}
                          onClick={handlers.testConnection}
                          className={classes.marginLeft}
                        >
                          Test Connection
                        </Button>
                        <Field
                          name="repository.connection_status"
                          subscription={{ value: true, error: true }}
                        >
                          {({ input: { value }, meta: { error } }) =>
                            !!error && <FormHelperText error>{error}</FormHelperText>
                          }
                        </Field>
                      </div>
                    </Grid>
                  </FormCondition>
                </Grid>
              </form>
            </React.Fragment>
          )
        }}
      </TestSourceForm>
    </div>
  )
}
Create.propTypes = {
  classes: PropTypes.object.isRequired,
  match: PropTypes.shape({
    params: PropTypes.shape({
      projectId: PropTypes.string.isRequired,
      testSourceId: PropTypes.string,
    }).isRequired,
  }).isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired,
  }).isRequired,
}

export default withStyles(styles)(Create)
