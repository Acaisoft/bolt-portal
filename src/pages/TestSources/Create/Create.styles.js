export default ({ spacing }) => ({
  marginLeft: {
    marginLeft: spacing.unit,
  },
})
