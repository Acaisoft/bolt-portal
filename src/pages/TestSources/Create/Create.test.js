import React from 'react'
import { shallow } from 'enzyme'

import { Create } from './Create'
import { mocks } from '~utils/tests'

const initCreate = overrides => {
  const mockProps = {
    match: {
      path: '/test-sources/create',
      url: '/test-sources/create',
      params: {
        projectId: 'mockedProjectId',
        testSourceId: 'mockedTestSourceId',
      },
    },
    history: {
      push: jest.fn(),
    },
  }
  const wrapper = shallow(
    <Create classes={mocks.ClassesProxy} {...mockProps} {...overrides} />
  )
  return { wrapper }
}

describe('page: Create', () => {
  describe('rendering', () => {
    it('should render without crashing', () => {
      const { wrapper } = initCreate()
      expect(wrapper).toBeTruthy()
    })
    it('should as expected', () => {
      const { wrapper } = initCreate()
      expect(wrapper).toMatchSnapshot()
    })
  })
})
