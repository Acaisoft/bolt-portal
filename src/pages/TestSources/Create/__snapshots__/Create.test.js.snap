// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`page: Create rendering should as expected 1`] = `
<div>
  <withApollo(Apollo(TestSourceForm))
    mode="create"
    onSubmit={[Function]}
    projectId="mockedProjectId"
  >
    <Component />
  </withApollo(Apollo(TestSourceForm))>
</div>
`;
